### PR TITLE
build v1.0

### PR DESCRIPTION
**Pull request by dawilles**

### CODE REVIEW

File: heroes.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+const Heroes = (() => {
```
Comment (zem-aleks) [inline heroes.js]: Please, NEVER wrap classes into module. Class is a type

File: render.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		mana: ".knight-mana",
+	},
+	wizard: {
+		health: ".mag-health",
+		energy: ".mag-energy",
+		mana: ".mag-mana",
+	},
+};
+
+const Render = (() => {
```
Comment (zem-aleks) [inline render.js]: It's no a class, should start from small letter

File: render.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+const selectors = {
```
Comment (zem-aleks) [inline render.js]: It's a global const. Sometimes it's okay to left it as it's now, but to clearly explain idea it's better to make SELECTORS and attributes also with big letters: KNIGHT: { HEALTH: "" }

File: ai.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+const IA = (() => {
```
Comment (zem-aleks) [inline ai.js]: not big letters and it's not clear what this means. Maybe you meant ia as artificial intelligence? Anyway, it's better to explain that it's IA belong to heroes or game or some comments :)

File: game.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+				: new Heroes.Magician(100, 100, 100);
+		computer =
+			choice === "knight"
+				? new Heroes.Magician(100, 100, 100)
+				: new Heroes.Paladin(100, 100, 100);
+		IA.setOpponent(player);
+	};
+
+	const turn = (action) => {
+		let playerDamage = player[action]();
```
Comment (zem-aleks) [inline game.js]: Wy do you use let instead of const?

File: game.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
+const Game = (() => {
+	let player;
+	let computer;
+	let playerChoice;
+
+	const start = (choice) => {
+		playerChoice = choice;
+		player =
+			choice === "knight"
```
Comment (zem-aleks) [inline game.js]: It's better to make a switch/case by choice

File: game.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+const Game = (() => {
```
Comment (zem-aleks) [inline game.js]: Heroes are created with classes. It's better to continue style of code and also make it as a class. Only classes start with big letter. Module is a variable, not a class!!!

File: game.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		playerChoice = choice;
+		player =
+			choice === "knight"
+				? new Heroes.Paladin(100, 100, 100)
+				: new Heroes.Magician(100, 100, 100);
+		computer =
+			choice === "knight"
+				? new Heroes.Magician(100, 100, 100)
+				: new Heroes.Paladin(100, 100, 100);
+		IA.setOpponent(player);
```
Comment (zem-aleks) [inline game.js]: IA is a game dependency. Must provided as constructor parameter or as a function parameter. If ia.js file is not connected or you will decide to use another ia implementation, game either will be broken or must be refactored

File: game.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+		if (action === "attack" && computerMove !== "defend") {
+			computer.hurt(playerDamage);
+		}
+
+		if (computerMove === "attack" && action !== "defend") {
+			player.hurt(computerDamage);
+		}
+
+		Render.updateUI(player, computer, playerChoice);
```
Comment (zem-aleks) [inline game.js]: Game only should wok with a game logic and has no options to call renders. You can create some events inside your game and app can listen these events and do some rendering, but no rendering from the game!

File: ai.js (hunk 4 above)
Comment (zem-aleks) [reply ai.js]: And it's better to make it as class, because it's abstraction that can be extended later

**Review by zem-aleks — CHANGES_REQUESTED** [review]

File: app.js [hunk 10]
```diff
@@ -1,46 +1,86 @@
...
+					onTurnFinished: () => {},
+				},
+			});
+			initGameInterface(game);
+			initGameEvents(game);
+		};
 
-	const showElement = (selector) => {
-		document.querySelector(selector).style.display = "block";
+		document
```
Comment (zem-aleks) [inline app.js]: 1 event listener for 2 controls

File: gameAi.js [hunk 11]
```diff
@@ -1,32 +1,20 @@
...
-	let opponent;
-	let self;
-	const setOpponent = (opponentHero, selfHero) => {
-		opponent = opponentHero;
-		self = selfHero;
-	};
-
-	const getMove = () => {
+class GameAi {
+	getMove(opponent) {
```
Comment (zem-aleks) [inline gameAi.js]: could be game as a main parameter. Based on game state you can make all decisions

**Review by zem-aleks — CHANGES_REQUESTED** [review]

File: app.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
+const app = (() => {
+	const init = () => {
+		const render = new Render();
+		const gameAi = new GameAi();
+		const createGame = (isKnight) => {
```
Comment (zem-aleks) [inline app.js]: move this function outside of init function

File: app.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+		};
+
+		setupEventListener(".form-knight", "submit", (event) => {
+			event.preventDefault();
+			const action = getCheckedValue(".form-knight", "knight-action");
+			const computerAction = gameAi.getMove(game);
+			game.turn(action, computerAction, render);
+		});
+
+		setupEventListener(".form-mag", "submit", (event) => {
```
Comment (zem-aleks) [inline app.js]: code duplication! DRY :)

File: gameAi.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
+class GameAi {
+	getMove(game) {
+		const opponent = game.computerHero;
+		if (opponent.health < 20) {
+			return "defend";
+		} else if (opponent.mana < 20 || opponent.energy < 20) {
```
Comment (zem-aleks) [inline gameAi.js]: don't make "else if", if you have a return in the previous statement. It's enough to make new "if". 
Such statements are easier for understanding

File: gameModule.js [hunk 15]
```diff
@@ -0,0 +1,68 @@
+class Game {
+	constructor(playerHero, computerHero, options) {
+		this.options = {
+			events: {
```
Comment (zem-aleks) [inline gameModule.js]: Please move it to DEFAULT_EVENTS constant

and this function can be used: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

**Review by zem-aleks — CHANGES_REQUESTED** [review]

File: gameAi.js (hunk 11 above)
Comment (dawilles) [reply gameAi.js]: done

File: gameAi.js (hunk 14 above)
Comment (dawilles) [reply gameAi.js]: done

File: app.js (hunk 13 above)
Comment (dawilles) [reply app.js]: done

File: gameModule.js (hunk 15 above)
Comment (dawilles) [reply gameModule.js]: done

File: app.js (hunk 12 above)
Comment (dawilles) [reply app.js]: done

**Pull request merged**